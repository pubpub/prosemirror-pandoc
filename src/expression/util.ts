import { Expr } from "./types";

export const exprAcceptsMultiple = (expr: Expr): boolean => {
    if (expr.type === "identifier") {
        return false;
    } else if (expr.type === "sequence") {
        return true;
    } else if (expr.type === "oneOrMore") {
        return true;
    } else if (expr.type === "zeroOrMore") {
        return true;
    } else if (expr.type === "range") {
        return expr.upperBound === null || expr.upperBound > 1;
    } else if (expr.type === "choice") {
        return expr.children.some((child) => exprAcceptsMultiple(child));
    }
};

export const exprWillAlwaysMatchSingleIdentifier = (expr: Expr, id: string) => {
    if (expr.type === "identifier") {
        return expr.identifier === id;
    } else if (expr.type === "sequence") {
        return false;
    } else if (expr.type === "choice") {
        return expr.children.some((child) =>
            exprWillAlwaysMatchSingleIdentifier(child, id)
        );
    } else if (expr.type === "range") {
        return (
            expr.lowerBound === 1 &&
            exprWillAlwaysMatchSingleIdentifier(expr.child, id)
        );
    } else {
        return exprWillAlwaysMatchSingleIdentifier(expr.child, id);
    }
};
