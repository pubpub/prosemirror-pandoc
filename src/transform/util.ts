import { Attr, Str, Space } from "../types";

export const createAttr = (
    identifier: string,
    classes?: string[],
    properties?: Record<string, any>
): Attr => {
    return { identifier, classes, properties };
};

export const textFromStrSpace = (nodes: (Str | Space)[]) => {
    let text = "";
    for (const entry of nodes) {
        if (entry.type === "Str") {
            text = text + entry.content;
        } else {
            text = text + " ";
        }
    }
    return text;
};

export const intersperse = (
    arr: any[],
    intersperseFn: (index?: number) => any
): any[] =>
    (Array.isArray(arr) ? arr : [arr]).reduce(
        (accumulated: any[], next: any, index: number): any[] => {
            const added: any[] = [next];
            if (index !== (Array.isArray(arr) ? arr : [arr]).length - 1) {
                added.push(intersperseFn(index));
            }
            return [...accumulated, ...added];
        },
        []
    );

export const textToStrSpace = (text: string): (Str | Space)[] =>
    intersperse(
        text.split(" ").map((word) => ({ type: "Str", content: word })),
        () => ({ type: "Space" })
    );

export const asArray = <T>(item: T | T[]): T[] => {
    return Array.isArray(item) ? item : [item];
};

export const asNode = <T>(item: T | T[]): T => {
    return Array.isArray(item) ? item[0] : item;
};

export const flatten = <T>(input: any): T[] => {
    if (!Array.isArray(input)) {
        return [input];
    }
    return input.reduce((arr: T[], next: T | T[]) => {
        if (Array.isArray(next)) {
            return [...arr, ...flatten(next)];
        }
        return [...arr, next];
    }, [] as T[]) as T[];
};

export const getQuoteChar = (
    single: boolean,
    opening: boolean,
    smart: boolean
) => {
    if (smart) {
        if (single) {
            return opening ? "‘" : "’";
        } else {
            return opening ? "“" : "”";
        }
    } else {
        return single ? "'" : '"';
    }
};
