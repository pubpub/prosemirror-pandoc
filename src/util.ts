import { execSync, spawnSync } from "child_process";

import { parsePandocJson } from "./parse";
import { RuleSet } from "./transform/ruleset";
import { fromPandoc } from "./transform/fromPandoc/fromPandoc";
import { PandocJson } from "./types";

const MAX_BUFFER = 0 * 1024 * 1024;

export const callPandoc = (
    source: string,
    inputFormat: string,
    outputFormat: string = "json",
    extraArgs: string[] = []
) => {
    return spawnSync(
        "pandoc",
        ["-f", inputFormat, "-t", outputFormat, "--quiet", ...extraArgs],
        { input: source, maxBuffer: MAX_BUFFER }
    ).stdout.toString();
};

export const callPandocWithFile = (
    sourcePath: string,
    outputFormat: string = "json",
    inputFormat: string = null,
    extraArgs: string[] = []
) => {
    const extraArgsString = extraArgs.join(" ");
    const inputFormatString = inputFormat ? `-f ${inputFormat}` : "";
    return execSync(
        `pandoc ${sourcePath} ${inputFormatString} -t ${outputFormat} ${extraArgsString}`,
        { maxBuffer: MAX_BUFFER }
    ).toString();
};

export const loadAndTransformFromPandoc = (
    sourcePath: string,
    rules: RuleSet<any>
) => {
    const pandocResult = callPandocWithFile(sourcePath);
    let json: PandocJson;
    try {
        json = JSON.parse(pandocResult);
    } catch (err) {
        if (pandocResult) {
            console.error(`Couldn't parse Pandoc result: ${pandocResult}`);
        } else {
            console.error(err);
        }
    }
    const pandocAst = parsePandocJson(json);
    return fromPandoc(pandocAst, rules).asNode();
};
