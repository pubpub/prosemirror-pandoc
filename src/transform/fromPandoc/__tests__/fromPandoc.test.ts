/* global describe, it, expect */
import rules from "../../../example/rules";

import { fromPandoc } from "..";
import { createAttr } from "../../util";
import {
    Header,
    OrderedList,
    BulletList,
    Div,
    Para,
    Plain,
    DefinitionList,
    Link,
} from "../../../types";

describe("fromPandoc", () => {
    it("transforms a Null into an empty array", () => {
        expect(fromPandoc({ type: "Null" }, rules)).toEqual([]);
    });

    it("transforms a simple string", () => {
        expect(
            fromPandoc(
                [
                    { type: "Str", content: "Hello" },
                    { type: "Space" },
                    { type: "Str", content: "world!" },
                ],
                rules
            ).asNode()
        ).toEqual({ type: "text", text: "Hello world!" });
    });

    it("transforms multiple paragraphs", () => {
        expect(
            fromPandoc(
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Hello" },
                            { type: "Space" },
                            { type: "Str", content: "dad!" },
                        ],
                    },
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Hello" },
                            { type: "Space" },
                            { type: "Str", content: "mom!" },
                        ],
                    },
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Ch-ch-ch-ch-ch-cherry" },
                            { type: "Space" },
                            { type: "Str", content: "bomb!" },
                        ],
                    },
                ],
                rules
            )
        ).toEqual([
            {
                type: "paragraph",
                content: [{ type: "text", text: "Hello dad!" }],
            },
            {
                type: "paragraph",
                content: [{ type: "text", text: "Hello mom!" }],
            },
            {
                type: "paragraph",
                content: [
                    { type: "text", text: "Ch-ch-ch-ch-ch-cherry bomb!" },
                ],
            },
        ]);
    });

    it("transforms a Para, Plain, or Div into a paragraph", () => {
        const blockOne: Para = {
            type: "Para",
            content: [
                { type: "Str", content: "Hello" },
                { type: "Space" },
                { type: "Str", content: "world!" },
            ],
        };
        const blockTwo: Plain = {
            type: "Plain",
            content: [
                { type: "Str", content: "Hello" },
                { type: "Space" },
                { type: "Str", content: "world!" },
            ],
        };
        const blockThree: Div = {
            type: "Div",
            attr: createAttr(""),
            content: [
                {
                    type: "Para",
                    content: [
                        { type: "Str", content: "Hello" },
                        { type: "Space" },
                        { type: "Str", content: "world!" },
                    ],
                },
            ],
        };
        [blockOne, blockTwo, blockThree].forEach(block =>
            expect(fromPandoc(block, rules).asNode()).toEqual({
                type: "paragraph",
                content: [{ type: "text", text: "Hello world!" }],
            })
        );
    });

    it("transforms a LineBlock into a paragraph", () => {
        expect(
            fromPandoc(
                {
                    type: "LineBlock",
                    content: [
                        [
                            { type: "Str", content: "Hello" },
                            { type: "Space" },
                            { type: "Str", content: "world!" },
                        ],
                        [
                            { type: "Str", content: "I'm" },
                            { type: "Space" },
                            { type: "Str", content: "testing!" },
                        ],
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "paragraph",
            content: [
                { type: "text", text: "Hello world!" },
                { type: "hard_break" },
                { type: "text", text: "I'm testing!" },
            ],
        });
    });

    it("transforms a CodeBlock into a code_block", () => {
        expect(
            fromPandoc(
                {
                    type: "CodeBlock",
                    content: "hello_world()",
                    attr: createAttr(""),
                },
                rules
            ).asNode()
        ).toEqual({ type: "code_block", text: "hello_world()" });
    });

    it("transforms a BlockQuote into a blockquote", () => {
        expect(
            fromPandoc(
                {
                    type: "BlockQuote",
                    content: [
                        {
                            type: "Para",
                            content: [
                                { type: "Str", content: "Lorem" },
                                { type: "Space" },
                                { type: "Str", content: "Ipsum?" },
                            ],
                        },
                        {
                            type: "Para",
                            content: [
                                { type: "Str", content: "More" },
                                { type: "Space" },
                                { type: "Str", content: "like" },
                                { type: "Space" },
                                { type: "Str", content: "Borem" },
                                { type: "Space" },
                                { type: "Str", content: "Ipsum!" },
                            ],
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "blockquote",
            content: [
                {
                    type: "paragraph",
                    content: [{ type: "text", text: "Lorem Ipsum?" }],
                },
                {
                    type: "paragraph",
                    content: [
                        { type: "text", text: "More like Borem Ipsum!" },
                    ],
                },
            ],
        });
    });

    it("transforms a Header into a heading", () => {
        const header: Header = {
            type: "Header",
            attr: createAttr("lesson-one", [], {}),
            level: 1,
            content: [
                { type: "Str", content: "Lesson" },
                { type: "Space" },
                { type: "Str", content: "One" },
            ],
        };
        expect(fromPandoc(header, rules).asNode()).toEqual({
            type: "heading",
            attrs: {
                level: 1,
                id: "lesson-one",
            },
            content: [{ type: "text", text: "Lesson One" }],
        });
    });

    it("transforms a HorizontalRule into a horizontal_rule", () => {
        expect(fromPandoc({ type: "HorizontalRule" }, rules).asNode()).toEqual({
            type: "horizontal_rule",
        });
    });

    it("transforms an OrderedList into an ordered_list", () => {
        const input: OrderedList = {
            type: "OrderedList",
            listAttributes: {
                startNumber: 2,
                listNumberDelim: "DefaultDelim",
                listNumberStyle: "DefaultStyle",
            },
            content: [
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "One" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Two" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Three" },
                        ],
                    },
                ],
            ],
        };
        const expectedOutput = {
            type: "ordered_list",
            attrs: {
                order: 2,
            },
            content: [
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item One" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Two" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Three" }],
                        },
                    ],
                },
            ],
        };
        expect(fromPandoc(input, rules).asNode()).toEqual(expectedOutput);
    });

    it("transforms a DefinitionList into a bullet_list", () => {
        const input: DefinitionList = {
            type: "DefinitionList",
            entries: [
                {
                    term: [
                        { type: "Str", content: "A" },
                        { type: "Space" },
                        { type: "Str", content: "term" },
                    ],
                    definitions: [
                        [
                            {
                                type: "Para",
                                content: [
                                    { type: "Str", content: "Definition" },
                                    { type: "Space" },
                                    { type: "Str", content: "One" },
                                ],
                            },
                            {
                                type: "Para",
                                content: [
                                    { type: "Str", content: "Paragraph" },
                                    { type: "Space" },
                                    { type: "Str", content: "B" },
                                ],
                            },
                        ],
                        [
                            {
                                type: "BlockQuote",
                                content: [
                                    {
                                        type: "Para",
                                        content: [
                                            {
                                                type: "Str",
                                                content: "Inside",
                                            },
                                            { type: "Space" },
                                            {
                                                type: "Str",
                                                content: "Blockquote",
                                            },
                                        ],
                                    },
                                ],
                            },
                            {
                                type: "Para",
                                content: [
                                    { type: "Str", content: "Paragraph" },
                                    { type: "Space" },
                                    { type: "Str", content: "D" },
                                ],
                            },
                        ],
                    ],
                },
                {
                    term: [{ type: "Str", content: "Mercifully" }],
                    definitions: [
                        [
                            {
                                type: "Para",
                                content: [{ type: "Str", content: "simple." }],
                            },
                        ],
                    ],
                },
            ],
        };
        expect(fromPandoc(input, rules)).toMatchSnapshot();
    });

    it("transforms an BulletList into an bullet_list", () => {
        const input: BulletList = {
            type: "BulletList",
            content: [
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "One" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Two" },
                        ],
                    },
                ],
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "Three" },
                        ],
                    },
                ],
            ],
        };
        const expectedOutput = {
            type: "bullet_list",
            attrs: {},
            content: [
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item One" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Two" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item Three" }],
                        },
                    ],
                },
            ],
        };
        expect(fromPandoc(input, rules).asNode()).toEqual(expectedOutput);
    });

    it("uses the processListItem argument of listTransformer to process list items", () => {
        // The PubPub prosemirror schema demands that the first element in a list_item
        // be a paragraph. We enforce that using a processListItem function, which we test here.
        const input: BulletList = {
            type: "BulletList",
            content: [
                [
                    {
                        type: "Para",
                        content: [
                            { type: "Str", content: "Item" },
                            { type: "Space" },
                            { type: "Str", content: "One" },
                        ],
                    },
                ],
                [
                    {
                        type: "BlockQuote",
                        content: [
                            {
                                type: "Para",
                                content: [
                                    { type: "Str", content: "Item" },
                                    { type: "Space" },
                                    { type: "Str", content: "Two" },
                                ],
                            },
                        ],
                    },
                ],
            ],
        };
        const expectedOutput = {
            type: "bullet_list",
            attrs: {},
            content: [
                {
                    type: "list_item",
                    content: [
                        {
                            type: "paragraph",
                            content: [{ type: "text", text: "Item One" }],
                        },
                    ],
                },
                {
                    type: "list_item",
                    content: [
                        // This is the added paragraph we expect to see.
                        { type: "paragraph", content: [] },
                        {
                            type: "blockquote",
                            content: [
                                {
                                    type: "paragraph",
                                    content: [
                                        { type: "text", text: "Item Two" },
                                    ],
                                },
                            ],
                        },
                    ],
                },
            ],
        };
        expect(fromPandoc(input, rules).asNode()).toEqual(expectedOutput);
    });

    it("uses the pandocPassThroughTransformer to ignore SmallCaps", () => {
        expect(
            fromPandoc(
                {
                    type: "Para",
                    content: [
                        {
                            type: "Str",
                            content: "Hello,",
                        },
                        { type: "Space" },
                        {
                            type: "SmallCaps",
                            content: [
                                { type: "Str", content: "small" },
                                { type: "Space" },
                                { type: "Str", content: "caps" },
                            ],
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "paragraph",
            content: [
                { type: "text", text: "Hello, " },
                { type: "text", text: "small caps" },
            ],
        });
    });

    it("performs a simple transformation from nodes to Prosemirror marks", () => {
        expect(
            fromPandoc(
                {
                    type: "Para",
                    content: [
                        {
                            type: "Strong",
                            content: [
                                { type: "Str", content: "Hello" },
                                { type: "Space" },
                                { type: "Str", content: "world!" },
                            ],
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "paragraph",
            content: [
                {
                    type: "text",
                    marks: [{ type: "strong" }],
                    text: "Hello world!",
                },
            ],
        });
    });

    it("transforms nested inline nodes into multiple Prosemirror marks", () => {
        expect(
            fromPandoc(
                {
                    type: "Para",
                    content: [
                        {
                            type: "Emph",
                            content: [
                                {
                                    type: "Strong",
                                    content: [
                                        { type: "Str", content: "Hello" },
                                        { type: "Space" },
                                        { type: "Str", content: "world!" },
                                    ],
                                },
                            ],
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "paragraph",
            content: [
                {
                    type: "text",
                    marks: [{ type: "strong" }, { type: "em" }],
                    text: "Hello world!",
                },
            ],
        });
    });

    it("ignores doubly-nested marks", () => {
        expect(
            fromPandoc(
                {
                    type: "Para",
                    content: [
                        {
                            type: "Emph",
                            content: [
                                {
                                    type: "Strong",
                                    content: [
                                        {
                                            type: "Emph",
                                            content: [
                                                {
                                                    type: "Str",
                                                    content: "Hello",
                                                },
                                                { type: "Space" },
                                                {
                                                    type: "Str",
                                                    content: "world!",
                                                },
                                            ],
                                        },
                                        {
                                            type: "Str",
                                            content: "Hello",
                                        },
                                        { type: "Space" },
                                        {
                                            type: "Str",
                                            content: "again!",
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "paragraph",
            content: [
                {
                    type: "text",
                    marks: [{ type: "em" }, { type: "strong" }],
                    text: "Hello world!",
                },
                {
                    type: "text",
                    marks: [{ type: "strong" }, { type: "em" }],
                    text: "Hello again!",
                },
            ],
        });
    });

    it("handles a complicated nodes -> marks nesting situation", () => {
        expect(
            fromPandoc(
                [
                    {
                        type: "Para",
                        content: [
                            {
                                type: "Strong",
                                content: [
                                    {
                                        type: "Subscript",
                                        content: [
                                            {
                                                type: "Str",
                                                content: "Hello",
                                            },
                                        ],
                                    },
                                    { type: "Space" },
                                    { type: "Str", content: "world!" },
                                ],
                            },
                        ],
                    },
                    {
                        type: "BlockQuote",
                        content: [
                            {
                                type: "Para",
                                content: [
                                    {
                                        type: "Strikeout",
                                        content: [
                                            { type: "Str", content: "What's" },
                                            { type: "Space" },
                                            {
                                                type: "Emph",
                                                content: [
                                                    {
                                                        type: "Str",
                                                        content: "up?",
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                ],
                rules
            ).asArray()
        ).toEqual([
            {
                type: "paragraph",
                content: [
                    {
                        type: "text",
                        marks: [{ type: "sub" }, { type: "strong" }],
                        text: "Hello",
                    },
                    {
                        type: "text",
                        marks: [{ type: "strong" }],
                        text: " world!",
                    },
                ],
            },
            {
                type: "blockquote",
                content: [
                    {
                        type: "paragraph",
                        content: [
                            {
                                type: "text",
                                marks: [{ type: "strike" }],
                                text: "What's ",
                            },
                            {
                                type: "text",
                                marks: [{ type: "em" }, { type: "strike" }],
                                text: "up?",
                            },
                        ],
                    },
                ],
            },
        ]);
    });

    it("gracefully handles marks that cannot be applied", () => {
        expect(
            fromPandoc(
                // This is incorrectly typed -- that's okay
                // @ts-ignore
                {
                    type: "Strong",
                    content: [{ type: "HorizontalRule" }],
                },
                rules
            ).asNode()
        ).toEqual({ type: "horizontal_rule" });
    });

    it("transforms Links", () => {
        const link: Link = {
            type: "Link",
            attr: createAttr(""),
            target: {
                url: "https://example.com",
                title: "Hi!",
            },
            content: [{ type: "Str", content: "Hey!" }],
        };
        expect(fromPandoc(link, rules).asNode()).toEqual({
            type: "text",
            marks: [
                {
                    type: "link",
                    attrs: { href: "https://example.com", title: "Hi!" },
                },
            ],
            text: "Hey!",
        });
    });

    it("handles an Image with a caption that needs to be converted to HTML by Pandoc", () => {
        expect(
            fromPandoc(
                {
                    type: "Image",
                    target: {
                        url: "https://pubpub.org/logo.png",
                        title: "The PubPub logo",
                    },
                    attr: createAttr(""),
                    content: [
                        {
                            type: "Strong",
                            content: [{ type: "Str", content: "Very" }],
                        },
                        { type: "LineBreak" },
                        { type: "Str", content: "cool." },
                    ],
                },
                rules
            ).asNode()
        ).toEqual({
            type: "image",
            attrs: {
                caption: "<p><strong>Very</strong><br />\ncool.</p>\n",
                url: "https://pubpub.org/logo.png",
            },
        });
    });

    it("transforms a Table", () => {
        expect(
            fromPandoc(
                {
                    caption: [{ type: "Str", content: "Caption!" }],
                    type: "Table",
                    alignments: ["AlignDefault", "AlignDefault"],
                    columnWidths: [0, 0],
                    headers: [
                        [
                            {
                                type: "Plain",
                                content: [{ type: "Str", content: "Apple" }],
                            },
                        ],
                        [
                            {
                                type: "Plain",
                                content: [{ type: "Str", content: "Rating" }],
                            },
                        ],
                    ],
                    cells: [
                        [
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "RedDelicious",
                                        },
                                    ],
                                },
                            ],
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "Terrible",
                                        },
                                    ],
                                },
                            ],
                        ],
                        [
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "GrannySmith",
                                        },
                                    ],
                                },
                            ],
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "Bad",
                                        },
                                    ],
                                },
                            ],
                        ],
                        [
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "Jazz",
                                        },
                                    ],
                                },
                            ],
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "PrettyGood",
                                        },
                                    ],
                                },
                            ],
                        ],
                        [
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "Honeycrisp",
                                        },
                                    ],
                                },
                            ],
                            [
                                {
                                    type: "Plain",
                                    content: [
                                        {
                                            type: "Str",
                                            content: "Perfect",
                                        },
                                    ],
                                },
                            ],
                        ],
                    ],
                },
                rules
            )
        ).toMatchSnapshot();
    });

    it("transforms Math (mathType=InlineMath) into an equation", () => {
        expect(
            fromPandoc(
                {
                    type: "Math",
                    mathType: "InlineMath",
                    content: "e^{i\\pi} = -1",
                },
                rules
            )
        ).toMatchSnapshot();
    });

    it("transforms Math (mathType=DisplayMath) into a block_equation", () => {
        expect(
            fromPandoc(
                {
                    type: "Math",
                    mathType: "DisplayMath",
                    content: "e^{i\\pi} = -1",
                },
                rules
            )
        ).toMatchSnapshot();
    });
});
