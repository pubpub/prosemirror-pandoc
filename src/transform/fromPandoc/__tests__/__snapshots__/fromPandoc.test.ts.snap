// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromPandoc transforms a DefinitionList into a bullet_list 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "marks": Array [
                  Object {
                    "type": "strong",
                  },
                ],
                "text": "A term: ",
                "type": "text",
              },
              Object {
                "text": "Definition One",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
          Object {
            "children": Array [
              Object {
                "text": "Paragraph B",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "text": "Inside Blockquote",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "blockquote",
          },
          Object {
            "children": Array [
              Object {
                "text": "Paragraph D",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list_item",
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "marks": Array [
                  Object {
                    "type": "strong",
                  },
                ],
                "text": "Mercifully: ",
                "type": "text",
              },
              Object {
                "text": "simple.",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "list_item",
      },
    ],
    "type": "bullet_list",
  },
]
`;
